tests/Functional/RetryDownloadFileTest.php

```diff
@@ -2,77 +2,104 @@
 
 namespace Keboola\S3ExtractorTest\Functional;
 
-use Aws\Command;
 use Aws\S3\Exception\S3Exception;
-use Aws\S3\S3Client;
 use Monolog\Handler\TestHandler;
 use Monolog\Logger;
-use Keboola\S3Extractor\DownloadFile;
-use PHPUnit\Framework\MockObject\MockObject;
+use Keboola\S3Extractor\S3AsyncDownloader;
 
 class RetryDownloadFileTest extends FunctionalTestCase
 {
     public function testRetrySuccess(): void
     {
+        $retryFile = 'retry-file.csv';
+        self::s3Client()->deleteObject([
+            'Bucket' => getenv(self::AWS_S3_BUCKET_ENV),
+            'Key' => $retryFile,
+        ]);
+
+        $handler = new TestHandler;
+        $tempPath = self::makeTempPath('retry-success');
         $consecutive = 0;
-        $client = $this->mockS3Client();
-        $client->method('getObject')
-            ->willReturnCallback(static function ($args) use (&$consecutive) {
-                $consecutive++;
-                if ($consecutive < 3) {
-                    throw new S3Exception(
-                        'Error executing "GetObject" on "foo"; AWS HTTP error: Server error: `GET bar` ' .
-                        'resulted in a `503 Slow Down` response:',
-                        new Command('dummy')
-                    );
-                }
 
-                file_put_contents($args['SaveAs'], 'dummy content');
-            });
+        $retryCallback = static function (array $fileParameters) use (&$consecutive, $tempPath) {
```

Comment:
Zajímavej způsob, jak testovat retries! Imho by to šlo snad i unittestovat tak, že bys do `CommandPool`u poslal mock `S3Client`, kterej by měl namockovanou metodu `S3Client::executeAsync(CommandInterface $command)`. Ta by mohla reagovat na každý `CommandInterface` jinak. Byla by to teda dost brutální kaskáda mocků, ale interface třídy `S3AsyncDownloader` by zůstal nezašpiněn, protože argument `$retryCallback` v konstruktoru je tam prostě navíc a nedá se bez čtení kódu úplně snadno pochopit (fakt mi chvíli trvalo, než jsem se dobral k tomu, že je jen pro testy).

Reply:
Namockovat by to určitě bylo lepší, založil jsem si na to issue https://github.com/keboola/aws-s3-extractor/issues/55 a až bude čas, zkusím to vytunit 👊

Reply:
jawohl